crates/holochain/src/conductor/conductor.rs

```diff
@@ -839,39 +839,59 @@ mod network_impls {
-                        .unwrap_or_default())
-                }
-                None => {
-                    let dna_hashes = self
-                        .spaces
-                        .get_from_spaces(|space| (*space.dna_hash).clone());
-                    let mut out = Vec::new();
-                    for dna_hash in dna_hashes {
-                        let peer_store = self
-                            .holochain_p2p
-                            .peer_store(dna_hash)
-                            .await
-                            .map_err(|err| ConductorApiError::CellError(err.into()))?;
-                        let all_peers = peer_store.get_all().await?;
-                        out.extend(all_peers);
-                    }
-                    Ok(out)
+            let dna_hashes = match maybe_dna_hashes {
+                Some(hashes) => hashes,
+                None => self
+                    .spaces
+                    .get_from_spaces(|space| (*space.dna_hash).clone()),
+            };
+
+            let mut out = HashSet::new();
+            for dna_hash in dna_hashes {
+                let peer_store = self
+                    .holochain_p2p
+                    .peer_store(dna_hash.clone())
+                    .await
+                    .map_err(|err| ConductorApiError::CellError(err.into()))?;
+                let all_peers = peer_store.get_all().await?;
+                out.extend(all_peers);
+            }
+            Ok(out.into_iter().collect())
+        }
+
+        /// Get signed agent info from the conductor for a given app
+        pub async fn get_app_agent_infos(
+            &self,
+            installed_app_id: &InstalledAppId,
+            maybe_dna_hashes: Option<Vec<DnaHash>>,
+        ) -> ConductorApiResult<Vec<Arc<AgentInfoSigned>>> {
+            // Get app info to know which DNAs belong to this app
+            let app_info = self.get_app_info(installed_app_id).await?.ok_or_else(|| {
+                ConductorApiError::other(format!("App not installed: {}", installed_app_id))
+            })?;
+
+            let mut app_dnas: HashSet<DnaHash> = HashSet::new();
+            for cell_infos in app_info.cell_info.values() {
+                for cell_info in cell_infos {
+                    let dna = match cell_info {
+                        CellInfo::Provisioned(cell) => cell.cell_id.dna_hash().clone(),
+                        CellInfo::Cloned(cell) => cell.cell_id.dna_hash().clone(),
+                        CellInfo::Stem(cell) => cell.original_dna_hash.clone(),
+                    };
+                    app_dnas.insert(dna);
                 }
             }
+
+            let hashes = match maybe_dna_hashes {
+                Some(mut dna_hashes) => {
+                    dna_hashes.retain(|h| app_dnas.contains(h));
+                    dna_hashes
+                }
+                None => app_dnas.into_iter().collect(),
+            };
+            self.get_agent_infos(Some(hashes)).await
```

Comment:
Nice, I'm liking this function now. It's a clear "filter down to what's in the app" and then call common functionality. Great!